vpc/service/e2e_test.go

```diff
@@ -0,0 +1,354 @@
+	if err != nil {
+		t.Fatalf("failed to start postgress container: %s", err)
+	}
+	testDB, err = dbContainer.Connect(ctx)
+	if err != nil {
+		t.Fatalf("failed to connect to test DB: %s", err)
+	}
+	// Set up tables
+	err = db.MigrateTo(ctx, testDB, 40, false)
+	if err != nil {
+		t.Fatalf("failed to set up tables: %s", err)
+	}
+}
+
+func shutdownDB(t *testing.T) {
+	err := dbContainer.Shutdown(context.Background())
+	if err != nil {
+		t.Fatalf("failed to clean up container: %s", err)
+	}
+	testDB.Close()
+}
+
+func waitForServer(t *testing.T, addr string) {
+	done := make(chan bool)
+	go func() {
+		for {
+			ctx := context.Background()
+			conn, err := grpc.DialContext(ctx, addr, grpc.WithInsecure())
+			if err != nil {
+				time.Sleep(100 * time.Millisecond)
+				continue
+			}
+			defer conn.Close()
+			healthClient := grpc_health_v1.NewHealthClient(conn)
+			healthcheckResponse, err := healthClient.Check(ctx, &grpc_health_v1.HealthCheckRequest{
+				Service: "com.netflix.titus.executor.vpc.TitusAgentVPCService",
+			})
+
+			if err == nil && healthcheckResponse.Status == grpc_health_v1.HealthCheckResponse_SERVING {
+				done <- true
+				return
+			}
+			time.Sleep(100 * time.Millisecond)
+		}
+	}()
+
+	select {
+	case <-done:
+		t.Logf("GPRC server is running and healthy")
+	case <-time.After(2 * time.Second):
+		t.Fatalf("GRPC server is still not healthy after 2s")
+	}
+}
+
+func runVpcService(ctx context.Context, t *testing.T, addr string,
+	mockSTS stsiface.STSAPI, mockEC2 ec2iface.EC2API) {
+	publicKey, privateKey, err := ed25519.GenerateKey(nil)
+	require.NoError(t, err)
+
+	key := vpcapi.PrivateKey{
+		Hostname:  "",
+		Generated: timestamppb.Now(),
+		Key: &vpcapi.PrivateKey_Ed25519Key_{
+			Ed25519Key: &vpcapi.PrivateKey_Ed25519Key{
+				Rfc8032Key: privateKey.Seed(),
+			},
+		},
+	}
+
+	// Insert public key
+	_, err = testDB.ExecContext(ctx,
+		"INSERT INTO trusted_public_keys(hostname, created_at, key, keytype) VALUES('test-host', now(), $1, 'ed25519')",
+		publicKey)
+	require.NoError(t, err)
+
+	vpcServiceConfig := &Config{
+		DBURL:                 dbContainer.DBURL(),
+		Key:                   key, // nolint:govet
+		MaxConcurrentRefresh:  10,
+		ReconcileInterval:     5 * time.Minute,
+		TLSConfig:             nil,
+		EnabledTaskLoops:      []string{},
+		EnabledLongLivedTasks: []string{},
+		WorkerRole:            "testWorkerRole",
+
+		TrunkNetworkInterfaceDescription:  trunkENIDescription,
+		BranchNetworkInterfaceDescription: branchENIDescription,
+
+		disableRouteCache: true,
+	}
+	vpcService, err := newVpcService(ctx, vpcServiceConfig)
+	require.NoError(t, err)
+
+	vpcService.ec2.NewSts = func(p client.ConfigProvider, cfgs ...*aws.Config) stsiface.STSAPI {
+		return mockSTS
+	}
+	vpcService.ec2.NewEC2 = func(p client.ConfigProvider, cfgs ...*aws.Config) ec2iface.EC2API {
+		return mockEC2
+	}
+
+	err = vpcService.run(ctx, addr)
+	require.NoError(t, err)
+}
+
+func TestProvisionInstanceV3(t *testing.T) {
```

Comment:
Confirmed this test really did run under buildkite and took 52 seconds.
Great job!

Reply:
It's a bit strange that both this test and `TableMetricsTest` took 50+ seconds. I thought the most time consuming step is to pull docker image. And both tests pull the same `postgres` docker image. I expected to see the test that runs later takes shorter time. Anyway, I will try figuring it out after I separate these tests to its own target.

Reply:
+1 to this!